### PR TITLE
Common FS events handling API

### LINKED ISSUE #726: Create a common fs events handling API for Linux/Windows/macOS
We now have fs event handling for all these platforms. We need find the common subset of APIs that can be supported on all of them and expose that through Streamly.FileSystem.Event module.

### PR DESCRIPTION
**Pull request by rnjtranjan**
Fixes #726

### CODE REVIEW

File: src/Streamly/FileSystem/Event.hs [hunk 1]
```diff
@@ -0,0 +1,109 @@
+-- |
+-- Module      : Streamly.FileSystem.Event
```
Comment (adithyaov) [inline src/Streamly/FileSystem/Event.hs:2]: Maybe put this module under `Internal` namespace for now?

File: src/Streamly/FileSystem/Event.hs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+import Streamly.Internal.Data.Array.Storable.Foreign (Array)
+
+#if defined(CABAL_OS_DARWIN)
+import qualified Streamly.Internal.FileSystem.Event.Darwin as Event
+#elif defined(CABAL_OS_LINUX)
+import qualified Streamly.Internal.FileSystem.Event.Linux as Event
+#elif defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.FileSystem.Event.Windows as Event
+#else
+#error "FS Events not supported on this platform
```
Comment (adithyaov) [inline src/Streamly/FileSystem/Event.hs:52]: Did you forget to close the double quote?

File: src/Streamly/FileSystem/Event.hs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    -- * Debugging
+    , showEvent
+    )
+where
+
+import Data.List.NonEmpty (NonEmpty)
+import Data.Word (Word8)
+import Streamly.Prelude (SerialT)
+import Streamly.Internal.Data.Array.Storable.Foreign (Array)
```
Comment (adithyaov) [inline src/Streamly/FileSystem/Event.hs:43]: Sort the imports

File: src/Streamly/FileSystem/Event.hs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+#elif defined(CABAL_OS_WINDOWS)
+import qualified Streamly.Internal.FileSystem.Event.Windows as Event
+#else
+#error "FS Events not supported on this platform
+#endif
+
+-------------------------------------------------------------------------------
+-- Main
+-------------------------------------------------------------------------------
+type Toggle = Event.Toggle
```
Comment (adithyaov) [inline src/Streamly/FileSystem/Event.hs:58]:
```suggestion

type Toggle = Event.Toggle
```

File: src/Streamly/FileSystem/Event.hs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+type Config = Event.Config
+type Event = Event.Event
+
+setRootMoved :: Toggle -> Config -> Config
+setRootMoved = Event.setRootMoved
+
+setAllEvents :: Toggle -> Config -> Config
+#if defined(CABAL_OS_DARWIN)
+setAllEvents = defaultConfig
```
Comment (adithyaov) [inline src/Streamly/FileSystem/Event.hs:67]: I think there will be a type mismatch here? I wonder why the CI for mac does not fail.
`defaultConfig :: Config`
`setAllEvents :: Toggle -> Config -> Config`

File: src/Streamly/FileSystem/Event.hs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+isCreated = Event.isCreated
+
+isDeleted :: Event -> Bool
+isDeleted = Event.isDeleted
+
+isMoved :: Event -> Bool
+#if defined(CABAL_OS_DARWIN)
+isMoved = Event.isMoved
+#else
+isMoved ev = Event.isMovedFrom ev || Event.isMovedTo ev
```
Comment (adithyaov) [inline src/Streamly/FileSystem/Event.hs:85]: Maybe create an `isMoved` in the modules for Linux and windows with the current definition? Then you can just say `isMoved = Event.isMoved`. You needn't follow it. This looks OK too.

File: src/Streamly/Internal/FileSystem/Event/Darwin.hs [hunk 7]
```diff
@@ -498,14 +499,15 @@ watchToStream (Watch handle _ _) =
...
--- non-existing path.  If a non-existing path is watched with 'setRootChanged'
+-- non-existing path.  If a non-existing path is watched with 'setRootMoved'
 -- then an 'isRootChanged' event is reported if the path is created later and
 -- the "path" field in the event is set to the dirname of the path rather than
 -- the full absolute path. This is the observed behavior on macOS 10.15.1.
 --
 -- @
 -- {-\# LANGUAGE MagicHash #-}
--- watchTreesWith ('setIgnoreSelf' 'On' . 'setRootChanged' 'On') [Array.fromCString\# "path"#]
+-- watchTreesWith ('setIgnoreSelf' 'On' . 'setRootMoved' 'On')
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Event/Darwin.hs:509]: This is a function application so it's OK for this to cross 80 columns.

File: src/Streamly/Internal/FileSystem/Event/Darwin.hs [hunk 8]
```diff
@@ -528,6 +530,23 @@ watchTreesWith f paths = S.bracket before after watchToStream
...
 
+-- | Start monitoring a list of file system paths for file system events with
+-- the supplied configuration. The paths could be files or directories.
+-- When the path is a directory, the whole directory tree under it is watched
+-- recursively. Monitoring starts from the current time onwards. The paths are
+-- specified as UTF-8 encoded 'Array' of 'Word8'.
+--
+-- @
+-- {-\# LANGUAGE MagicHash #-}
+-- watch True ('setIgnoreSelf' 'On' . 'setRootMoved' 'On')
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Event/Darwin.hs:541]: You can put this on the same line.

File: src/Streamly/Internal/FileSystem/Event/Darwin.hs [hunk 9]
```diff
@@ -528,6 +530,23 @@ watchTreesWith f paths = S.bracket before after watchToStream
...
+--  [Array.fromCString\# "path"#]
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Event/Darwin.hs:542]: You can put this on the same line.

File: src/Streamly/Internal/FileSystem/Event/Linux.hs [hunk 10]
```diff
@@ -839,35 +840,28 @@ watchToStream cfg wt@(Watch handle _) = do
...
--- current time onwards. The paths are specified as "/" separated 'Array' of
+-- current time onwards. The paths are specified as UTF-8 encoded 'Array' of
 -- 'Word8'.
 --
 -- @
--- watchPathsWith
---  ('setFollowSymLinks' On . 'setUnwatchMoved' Off)
+-- watch True
+--  ('setFollowSymLinks' On . 'setUnwatchMoved' Off) defaultConfig
 --  [Array.fromCString\# "dir"#]
```
Comment (adithyaov) [inline src/Streamly/Internal/FileSystem/Event/Linux.hs:855]:
```
watch 
    True
    ('setFollowSymLinks' On . 'setUnwatchMoved' Off)
    defaultConfig
    [Array.fromCString\# "dir"#]
```
Or,
```
watch 
    True ('setFollowSymLinks' On . 'setUnwatchMoved' Off) defaultConfig [Array.fromCString\# "dir"#]
```

File: streamly.cabal [hunk 11]
```diff
@@ -533,7 +534,7 @@ library
                      , transformers      >= 0.4   && < 0.6
 
                      , heaps             >= 0.3     && < 0.4
-
+                     , filepath          >= 1.4.1 && < 1.5
```
Comment (adithyaov) [inline streamly.cabal:537]: I know we have implementeing/describing proper file paths in our agenda. Maybe you can just use `String (FilePath)` for now. This is also fine but we'll be removing this sooner or later.

File: streamly.cabal [hunk 12]
```diff
@@ -398,6 +398,7 @@ library
                      , Streamly.FileSystem.IOVec
                      , Streamly.FileSystem.FDIO
                      , Streamly.FileSystem.FD
+                     , Streamly.FileSystem.Event
```
Comment (adithyaov) [inline streamly.cabal:401]: You should move this under `Internal` and expose it.

**Review by adithyaov — COMMENTED** [review]
If there are any benchmarks for `FileSystem.Event` you should copy them and write benchmarks for this as well (maybe in another PR).

I did not check the documentation changes. Harendra will have the right context to do so.

File: src/Streamly/FileSystem/Event.hs (hunk 6 above)
Comment (rnjtranjan) [reply src/Streamly/FileSystem/Event.hs]: keeping as it is.

File: streamly.cabal (hunk 11 above)
Comment (rnjtranjan) [reply streamly.cabal:537]: currently, filepath package is used for path concatenation which is platform-independent.

**Comment by harendra-kumar** [comment]
Can you please rebase this?

**Comment by rnjtranjan** [comment]
Due to a lot of merge conflicts a new PR https://github.com/composewell/streamly/pull/1197

**Pull request closed without merging**